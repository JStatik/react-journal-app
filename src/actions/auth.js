import { firebase, googleAuthProvider } from '../firebase/firebaseConfig'
import types from "../types/types";
import popUp from '../helpers/popups';
import { finishLoading, startLoading } from './ui';

export const startLoginWithEmailPassword = ( email, password ) => {
    return ( dispatch ) => {
        dispatch( startLoading() );

        return firebase.auth().signInWithEmailAndPassword( email, password ).then( ( { user } ) => {
            dispatch( login( user.uid, user.displayName, user.photoURL ) );
        } ).catch( ( { message } ) => {
            console.log( `Error: ${ message }` );
            popUp( 'error', '#990000', 'Error', 'Alguno de los datos ingresados no es correcto.' );
        } ).finally( () => {
            dispatch( finishLoading() );
        } );
    }
};

export const startGoogleLogin = () => {
    return ( dispatch ) => {
        dispatch( startLoading() );

        firebase.auth().signInWithPopup( googleAuthProvider ).then( ( { user } ) => {
            dispatch( login( user.uid, user.displayName, user.photoURL ) );
        } ).catch( ( { message } ) => {
            console.log( `Error: ${ message }` );
            popUp( 'error', '#990000', 'Error', 'Alguno de los datos ingresados no es correcto.' );
        } ).finally( () => {
            dispatch( finishLoading() );
        } );
    }
};

export const startRegisterWithNameEmailPassword = ( name, email, password ) => {
    return ( dispatch ) => {
        dispatch( startLoading() );

        firebase.auth().createUserWithEmailAndPassword( email, password ).then( async( { user } ) => {
            await user.updateProfile( { 
                displayName: name 
            } );

            dispatch( login( user.uid, user.displayName ) );
        } ).catch( ( { message } ) => {
            console.log( `Error: ${ message }` );
            popUp( 'error', '#990000', 'Error', 'El email ingresado ya estÃ¡ en uso.' );
        } ).finally( () => {
            dispatch( finishLoading() );
        } );
    }
};

export const login = ( uid, displayName, photoURL ) => (
    {
        type: types.login,
        payload: {
            uid: uid,
            displayName: displayName,
            photoURL: photoURL
        }
    }
);

export const startLogout = () => {
    return async( dispatch ) => {
        dispatch( startLoading() );
        await firebase.auth().signOut();

        dispatch( logout() );
        dispatch( finishLoading() );
    };
};

export const logout = () => (
    {
        type: types.logout
    }
);
