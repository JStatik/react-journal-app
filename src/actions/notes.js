import Swal from 'sweetalert2';
import { bd } from '../firebase/firebaseConfig';
import types from '../types/types';
import popUp from '../helpers/popups';
import sendImageCloudinary from '../helpers/cloudinary';

export const startNewNote = () => {
    return async( dispatch, getState ) => {
        const { uid } = getState().auth;

        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        };

        try {
            const docRef = bd.collection( `${ uid }/journal/notes` );
            const noteRef = await docRef.add( newNote );
        
            dispatch( activeNote( noteRef.id, newNote ) );
            dispatch( addNewNote( noteRef.id, newNote ) );
        } catch( error ) {
            console.log( error );
        }      
    };
};

export const startLoadingNotes = () => {
    return async( dispatch, getState ) => {
        const notes = [];
        const { uid } = getState().auth;

        try {
            const notesSnap = await bd.collection( `${ uid }/journal/notes` ).get();

            notesSnap.forEach( note => {
                notes.push( {
                    ...note.data(),
                    id: note.id                
                } );
            } );

            dispatch( setNotes( notes ) );
        } catch( error ) {
            console.log( error );
        }       
    };
};

export const startSaveNote = ( note ) => {
    return async( dispatch, getState ) => {
        const { uid } = getState().auth;

        !note.url && delete note.url;

        const noteToFirestore = { ...note };
        delete noteToFirestore.id;

        try {
            await bd.doc( `${ uid }/journal/notes/${ note.id }` ).update( noteToFirestore );

            dispatch( refreshNote( note ) );
            popUp( 'success', '#005928 ', 'Ã‰xito', 'Los cambios se han efectuado correctamente.' );
        } catch( error ) {
            console.log( error );
        }
    };
};

export const startUploading = ( file ) => {
    return async( dispatch, getState ) => {
        const { active: note } = getState().notes;

        Swal.fire( { 
            hideClass: {
                popup: 'animate__animated animate__zoomOut'
            },
            willOpen: () => {
                Swal.showLoading();
            },
            showClass: {
                popup: 'animate__animated animate__zoomIn'
            },
            title: '<h6>Subiendo imagen</h6>',
            text: 'Por favor, espere...',
            width: '27rem'
        } );

        const urlImage = await sendImageCloudinary( file );
        note.url = urlImage;

        dispatch( startSaveNote( note ) );

        Swal.close();
    };
};

export const startDeleting = ( id ) => {
    return ( dispatch, getState ) => {
        const { uid } = getState().auth;

        bd.doc( `${ uid }/journal/notes/${ id }` ).delete().then( () => {
            dispatch( deleteNote( id ) );
        } ).catch( error => {
            console.log( error );
        } );

        Swal.close();
    };
};

export const activeNote = ( id, note ) => (
    {
        type: types.notesActive,
        payload: {
            ...note,
            id: id
        }
    }
);

export const addNewNote = ( id, note ) => (
    {
        type: types.notesAddNew,
        payload: {
            ...note,
            id: id
        }
    }
);

export const setNotes = ( notes ) => (
    {
        type: types.notesLoad,
        payload: notes
    }
);

export const refreshNote = ( note ) => (
    {
        type: types.notesUpdate,
        payload: note
    }
);

export const deleteNote = ( id ) => (
    {
        type: types.notesDelete,
        payload: id
    }
);

export const noteLogout = () => (
    {
        type: types.notesLogoutCleaning,
        payload: {
            notes: [],
            active: null
        }
    }
);
