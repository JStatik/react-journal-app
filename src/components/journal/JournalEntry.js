import React from 'react';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import moment from 'moment';
import 'moment/locale/es';
import { activeNote } from '../../actions/notes';

const JournalEntry = ( { id, date, title, body, url } ) => {
    const dispatch = useDispatch();
    const noteDate = moment( date );

    const handleEntryClick = () => {
        dispatch( activeNote( id, { date, title, body, url } ) );
    };

    return (
        <div className="journal__entry animate__animated animate__fadeIn" onClick={ handleEntryClick }>
            {
                url && <div className="journal__entry-picture" style={ { backgroundPosition: 'center center', backgroundSize: 'cover', backgroundImage: `url( ${ url } )` } }></div>
            }

            <div className="journal__entry-body">
                <p className="journal__entry-title">{ title }</p>
                <p className="journal__entry-content">{ body }</p>
            </div>

            <div className="journal__entry-date-box">
                <span>{ noteDate.format( 'dddd' ) }</span>
                <h4>{ noteDate.format( 'DD/MM/YY' ) }</h4>
            </div>
        </div>
    )
};

JournalEntry.propTypes = {
    id: PropTypes.string.isRequired,
    date: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    body: PropTypes.string.isRequired
};

export default JournalEntry;
