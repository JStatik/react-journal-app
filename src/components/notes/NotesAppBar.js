import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import moment from 'moment';
import 'moment/locale/es';
import { startSaveNote, startUploading } from '../../actions/notes';
import popUp from '../../helpers/popups';

const NotesAppBar = () => {
    const date = moment( new Date() );
    const dispatch = useDispatch();
    const { active: note } = useSelector( store => store.notes );

    const handleSave = () => {
        dispatch( startSaveNote( note ) );
    };

    const handlePicture = () => {
        document.getElementById( 'fileSelector' ).click();
    };

    const handleFile = ( event ) => {
        const file = event.target.files[ 0 ];
        const nameFile = file?.name;

        if( nameFile ) {
            if( nameFile.includes( '.jpg' ) || nameFile.includes( '.jpeg' ) || nameFile.includes( '.png' ) || nameFile.includes( '.svg' ) ) {
                dispatch( startUploading( file ) );
                document.getElementById( 'fileSelector' ).value = '';
            } else {
                popUp( 'error', '#990000', 'Error', 'El archivo seleccionado no es correcto. Elija una imagen.' );
                document.getElementById( 'fileSelector' ).value = '';
            }
        }
    };

    return (
        <div className="notes__app-bar">
            <span><span className="capitalize">{ date.format( 'dddd' ) }, </span>{ `${ date.format( 'DD' ) } de ${ date.format( 'MMMM' ) } del ${ date.format( 'YYYY' ) }` }</span>

            <input type="file" name="file" id="fileSelector" style={ { display: 'none' } } onChange={ handleFile }/>

            <div>
                <button className="btn" onClick={ handlePicture }>Imagen</button>
                <button className="btn" onClick={ handleSave }>Guardar</button>
            </div>
        </div>
    )
};

export default NotesAppBar;
