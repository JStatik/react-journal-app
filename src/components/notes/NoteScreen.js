import React, { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { activeNote, startDeleting } from '../../actions/notes';
import useForm from '../../hooks/useForm';
import NotesAppBar from './NotesAppBar';

const NoteScreen = () => {
    const dispatch = useDispatch();
    const { active: note } = useSelector( store => store.notes );
    const [ formValues, handleInputChange, reset ] = useForm( note );
    const { id, title, body } = formValues;
    const activeId = useRef( id );
    const activeUrl = useRef( note.url );

    useEffect( () => {
        if( note.id !== activeId.current ) {
            reset( note );
            activeId.current = note.id;
        }

        if( note.url !== activeUrl.current ) {
            reset( note );
            activeUrl.current = note.url;
        }
    }, [ note, reset ] );

    useEffect( () => {
        dispatch( activeNote( formValues.id, { ...formValues } ) );
    }, [ formValues, dispatch ] );

    const handleDelete = () => {
        dispatch( startDeleting( id ) );
    };

    return (
        <div className="notes__main-content">
            <NotesAppBar />

            <div className="notes__content">
                <input type="text" name="title" placeholder="Algún buen título" className="notes__title-input" autoComplete="off" value={ title } onChange={ handleInputChange } />
                <textarea name="body" placeholder="¿Que sucedio hoy?" className="notes__textarea" autoComplete="off" value={ body } onChange={ handleInputChange }></textarea>
                {
                    note.url && 
                    <div className="notes__image">
                        <img src={ `${ note.url }` } alt={ `${ note.title }` }/>
                    </div>
                }
            </div>

            <button className="btn btn-danger" onClick={ handleDelete }>Borrar</button>
        </div>
    )
};

export default NoteScreen;
