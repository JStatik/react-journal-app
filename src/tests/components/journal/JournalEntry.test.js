import React from 'react';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import JournalEntry from '../../../components/journal/JournalEntry';
import { activeNote } from '../../../actions/notes';

jest.mock( '../../../actions/notes', () => (
    {
        activeNote: jest.fn()
    }
) );

const middlewares = [ thunk ];
const mockStore = configureStore( middlewares );

const store = mockStore( {} );

store.dispatch = jest.fn();

describe( 'Pruebas en JournalEntry', () => {
    const note = {
        id: 'ABC',
        title: 'Titulo en test',
        body: 'Body en test',
        date: 0,
        url: 'https://photo.png'
    };

    const wrapper = mount(
        <Provider store={ store }>
            <JournalEntry { ...note } />
        </Provider>
    );

    test( 'Debe mostrar el componente correctamente', () => {
        expect( wrapper ).toMatchSnapshot();
    } );

    test( 'Debe ejecutar la accion activeNote()', () => {
        const { id, title, body, date, url } = note;

        wrapper.find( '.journal__entry' ).simulate( 'click' );

        expect( activeNote ).toHaveBeenCalledTimes( 1 );
        expect( activeNote ).toHaveBeenCalledWith( id, {
            title: title,
            body: body,
            date: date,
            url: url
        } );
    } );
} );
