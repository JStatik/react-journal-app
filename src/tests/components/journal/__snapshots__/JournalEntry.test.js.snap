// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en JournalEntry Debe mostrar el componente correctamente 1`] = `
<div
  className="journal__entry animate__animated animate__fadeIn"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url( https://photo.png )",
        "backgroundPosition": "center center",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Titulo en test
    </p>
    <p
      className="journal__entry-content"
    >
      Body en test
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      mi√©rcoles
    </span>
    <h4>
      31/12/69
    </h4>
  </div>
</div>
`;
